### PR TITLE
Add state, commands, events and rejections for realms.

### PR DESCRIPTION
**Pull request by wwajerowicz**
Fixes https://github.com/BlueBrain/nexus/issues/249

### CODE REVIEW

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package ch.epfl.bluebrain.nexus.iam.realms
+import cats.effect.Async
+import ch.epfl.bluebrain.nexus.iam.realms.RealmEvent._
+import ch.epfl.bluebrain.nexus.iam.realms.RealmRejection._
+import ch.epfl.bluebrain.nexus.iam.realms.RealmState._
+
+sealed trait RealmCommand extends Product with Serializable
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala]: Missing the following information
```scala
def instant: Instant
def identity: Identity
```

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmEvent.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package ch.epfl.bluebrain.nexus.iam.realms
+
+sealed trait RealmEvent extends Product with Serializable
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmEvent.scala]: Missing the following information
```scala
def instant: Instant
def identity: Identity
```

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmState.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package ch.epfl.bluebrain.nexus.iam.realms
+import ch.epfl.bluebrain.nexus.iam.realms.RealmEvent._
+
+sealed trait RealmState extends Product with Serializable
+
+object RealmState {
+
+  final case object Initial                                             extends RealmState
+  final case class Current(realm: Realm, rev: Int, deprecated: Boolean) extends RealmState
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmState.scala]: Missing the following information
```scala
def instant: Instant
def identity: Identity
```

Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmState.scala]: Probably we also want to store in the current state the `TokenId` and `java.security.PublicKey`, being the tokenId:

```scala
/**
  * A token unique identifier
  *
  * @param iss the token issuer
  * @param kid the KEY ID of the key
  */
final case class TokenId(iss: Uri, kid: String)

object TokenId {
  final implicit val tokenIdDecoder: Decoder[TokenId] = Decoder.forProduct2[TokenId, String, String]("iss", "kid") {
    case (iss, kid) => TokenId(iss, kid)
  }
}
```

You can find [here](https://github.com/BlueBrain/nexus-iam/blob/b2ce115abc1386311471a48a47ae10811bb0cb11/modules/service/src/main/scala/ch/epfl/bluebrain/nexus/iam/service/auth/CredentialsStoreActor.scala) more information about it.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+object RealmCommand {
+
+  final case class CreateRealm(realm: Realm)           extends RealmCommand
+  final case class UpdateRealm(realm: Realm, rev: Int) extends RealmCommand
+  final case class DeprecateRealm(rev: Int)            extends RealmCommand
+
+  def evaluate[F[_]](state: RealmState, command: RealmCommand)(
+      implicit F: Async[F]): F[Either[RealmRejection, RealmEvent]] = (state, command) match {
+    case (Initial, CreateRealm(realm))      => F.pure(Right(RealmCreated(realm)))
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala]: The realm is not always valid by default. Here you will have to attempt to [fetch the PublicKey](https://github.com/BlueBrain/nexus-iam/blob/b2ce115abc1386311471a48a47ae10811bb0cb11/modules/service/src/main/scala/ch/epfl/bluebrain/nexus/iam/service/auth/CredentialsStoreActor.scala#L61) and pass it to the state or reject when not it does not pass.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/Realm.scala [hunk 5]
```diff
@@ -0,0 +1,4 @@
+package ch.epfl.bluebrain.nexus.iam.realms
+import ch.epfl.bluebrain.nexus.rdf.Iri.Url
+
+case class Realm(label: String, openidConfiguration: Url, requiredScopes: Set[String])
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/Realm.scala]: add `final`

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import ch.epfl.bluebrain.nexus.iam.realms.RealmEvent._
+import ch.epfl.bluebrain.nexus.iam.realms.RealmRejection._
+import ch.epfl.bluebrain.nexus.iam.realms.RealmState._
+
+sealed trait RealmCommand extends Product with Serializable
+
+object RealmCommand {
+
+  final case class CreateRealm(realm: Realm)           extends RealmCommand
+  final case class UpdateRealm(realm: Realm, rev: Int) extends RealmCommand
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala]: `rev` should be a `Long` (to be consistent with what we have elsewhere).

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    case (Initial, _)                       => F.pure(Left(RealmDoesNotExist))
+    case (Current(_, _, _), CreateRealm(_)) => F.pure(Left(RealmAlreadyExists))
+    case (Current(_, currentRev, false), UpdateRealm(realm, updateRev)) if currentRev == updateRev =>
+      F.pure(Right(RealmUpdated(realm, currentRev + 1)))
+    case (Current(_, currentRev, false), UpdateRealm(_, updateRev)) =>
+      F.pure(Left(IncorrectRevision(currentRev, updateRev)))
+    case (Current(_, _, true), _) => F.pure(Left(RealmRejection.RealmDeprecated))
+    case (Current(_, currentRev, false), DeprecateRealm(updateRev)) if currentRev == updateRev =>
+      F.pure(Right(RealmDeprecated(currentRev + 1)))
+    case (Current(_, currentRev, false), DeprecateRealm(updateRev)) =>
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala]: the last two cases are identical

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      F.pure(Left(IncorrectRevision(currentRev, updateRev)))
+    case (Current(_, currentRev, false), DeprecateRealm(updateRev)) =>
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmCommand.scala]: Also needs to handle the attempt to deprecate twice the same realm.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmEvent.scala [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+object RealmEvent {
+
+  final case class RealmCreated(realm: Realm)           extends RealmEvent
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/RealmEvent.scala]: add also `rev: Long` to the `RealmCreated ` event. That's how we have it in KG and ADMIN

**Pull request closed without merging**